weather_mv/loader_pipeline/metrics.py

```diff
@@ -90,50 +93,77 @@ def wrapper(self, *args, **kwargs):
                 raise ValueError("Function is not a generator.")
 
         return wrapper
+
     return decorator
 
 
 class AddTimer(beam.DoFn):
-    """DoFn to add a empty time_dict per element in PCollection. This dict will stage_names as keys
-    and the time it took for that element in that stage."""
+    """DoFn to add a empty time_dict per element in PCollection. This dict will
+    stage_names as keys and the time it took for that element in that stage."""
+
     def process(self, element) -> t.Iterator[t.Any]:
-        time_dict = {}
+        time_dict = OrderedDict([
+            ("uri", element),
+            ("bucket", get_file_time(element)),
+            ("pickup", time.time()),
+        ])
         yield element, time_dict
 
 
 class AddMetrics(beam.DoFn):
-    """DoFn to add Element Processing Time metric to beam. Expects PCollection to contain a time_dict."""
+    """DoFn to add Element Processing Time metric to beam. Expects PCollection
+    to contain a time_dict."""
 
-    def __init__(self, asset_start_time_format: str = '%Y-%m-%dT%H:%M:%SZ'):
+    def __init__(self, asset_start_time_format: str = "%Y-%m-%dT%H:%M:%SZ"):
         super().__init__()
-        self.element_processing_time = metric.Metrics.distribution('Time', 'element_processing_time_ms')
-        self.data_latency_time = metric.Metrics.distribution('Time', 'data_latency_time_ms')
+        self.element_processing_time = metric.Metrics.distribution(
+            "Time", "element_processing_time_ms"
+        )
+        self.data_latency_time = metric.Metrics.distribution(
+            "Time", "data_latency_time_ms"
+        )
         self.asset_start_time_format = asset_start_time_format
 
     def process(self, element):
         try:
             if len(element) == 0:
                 raise ValueError("time_dict not found.")
             (_, asset_start_time), time_dict = element
-            if not isinstance(time_dict, dict):
+            if not isinstance(time_dict, OrderedDict):
                 raise ValueError("time_dict not found.")
 
-            # Adding element processing time.
-            total_time = 0
-            for stage_time in time_dict.values():
-                total_time += stage_time
+            uri = time_dict.pop("uri", _)
 
-            # Converting seconds to milli seconds.
-            self.element_processing_time.update(int(total_time * 1000))
+            time_dict_values = list(time_dict.values())
+            element_processing_time = (time_dict_values[-1] - time_dict_values[0])
```

Comment:
Can we add comment above this line; exactly which timestamps we are subtracting.

Reply:
Also let's add a PR description specifying what `element_processing_time` mean & `data_latency_time` mean.

Reply:
Done